frame/contracts/src/migration/v12.rs

```diff
@@ -0,0 +1,322 @@
+use sp_core::hexdisplay::HexDisplay;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+use sp_runtime::{traits::Zero, FixedPointNumber, FixedU128, Saturating};
+use sp_std::{marker::PhantomData, prelude::*};
+
+mod old {
+	use super::*;
+
+	#[derive(Encode, Decode, scale_info::TypeInfo, MaxEncodedLen)]
+	#[codec(mel_bound())]
+	#[scale_info(skip_type_params(T))]
+	pub struct OwnerInfo<T: Config> {
+		pub owner: AccountIdOf<T>,
+		#[codec(compact)]
+		pub deposit: BalanceOf<T>,
+		#[codec(compact)]
+		pub refcount: u64,
+	}
+
+	#[derive(Encode, Decode, scale_info::TypeInfo)]
+	#[codec(mel_bound())]
+	#[scale_info(skip_type_params(T))]
+	pub struct PrefabWasmModule {
+		#[codec(compact)]
+		pub instruction_weights_version: u32,
+		#[codec(compact)]
+		pub initial: u32,
+		#[codec(compact)]
+		pub maximum: u32,
+		pub code: Vec<u8>,
+		pub determinism: Determinism,
+	}
+
+	#[storage_alias]
+	pub type OwnerInfoOf<T: Config> = StorageMap<Pallet<T>, Identity, CodeHash<T>, OwnerInfo<T>>;
+
+	#[storage_alias]
+	pub type CodeStorage<T: Config> =
+		StorageMap<Pallet<T>, Identity, CodeHash<T>, PrefabWasmModule>;
+}
+
+#[derive(Encode, Decode, scale_info::TypeInfo, MaxEncodedLen)]
+#[codec(mel_bound())]
+#[scale_info(skip_type_params(T))]
+pub struct CodeInfo<T: Config> {
+	owner: AccountIdOf<T>,
+	#[codec(compact)]
+	deposit: BalanceOf<T>,
+	#[codec(compact)]
+	refcount: u64,
+	determinism: Determinism,
+}
+
+#[storage_alias]
+pub type CodeInfoOf<T: Config> = StorageMap<Pallet<T>, Twox64Concat, CodeHash<T>, CodeInfo<T>>;
+
+#[storage_alias]
+pub type PristineCode<T: Config> = StorageMap<Pallet<T>, Identity, CodeHash<T>, Vec<u8>>;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub fn store_old_dummy_code<T: Config>(len: usize, account: T::AccountId) {
+	use sp_runtime::traits::Hash;
+
+	let code = vec![42u8; len];
+	let hash = T::Hashing::hash(&code);
+	PristineCode::<T>::insert(hash, code.clone());
+
+	let module = old::PrefabWasmModule {
+		instruction_weights_version: Default::default(),
+		initial: Default::default(),
+		maximum: Default::default(),
+		code,
+		determinism: Determinism::Enforced,
+	};
+	old::CodeStorage::<T>::insert(hash, module);
+
+	let info = old::OwnerInfo { owner: account, deposit: u32::MAX.into(), refcount: u64::MAX };
+	old::OwnerInfoOf::<T>::insert(hash, info);
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, DefaultNoBound)]
+pub struct Migration<T: Config> {
+	last_code_hash: Option<CodeHash<T>>,
+	_phantom: PhantomData<T>,
+}
+
+impl<T: Config> MigrationStep for Migration<T> {
+	const VERSION: u16 = 12;
+
+	fn max_step_weight() -> Weight {
+		T::WeightInfo::v12_migration_step(T::MaxCodeLen::get())
+	}
+
+	fn step(&mut self) -> (IsFinished, Weight) {
+		let mut iter = if let Some(last_key) = self.last_code_hash.take() {
+			PristineCode::<T>::iter_from(PristineCode::<T>::hashed_key_for(last_key))
+		} else {
+			PristineCode::<T>::iter()
+		};
+		if let Some((hash, code)) = iter.next() {
```

Comment:
I think this will be a very expensive migration if we read all three storage items. IMHO we can just leave `PristineCode` alone and refund half of the deposit. It is not 100% correct but close enough. It also makes the code much simpler.

Reply:
And now that we have MBMs in place, is migration's large weight even a problem?

Reply:
Alright, re-ran the try-runtime for the last 4 migrations, got the following numbers 

 | migration |      ref_time | % of v12 | proof_size | % of v12 |
 |-----------|---------------|----------|------------|----------|
 | v9        |  174684909068 |     0.17 |   27728976 |     0.32 |
 | v10       |  572743124068 |     0.57 |   41084911 |     0.48 |
 | v11       |  574246502038 |     0.57 |   41092187 |     0.48 |
 | v12       | 1004688807116 |     1.00 |   85580551 |     1.00 |

So basically yeah, there is some room for optimization. Let's see how it goes if we iterate over OwnerInfos only, not reading the PristineCode at all.

Reply:
I am mainly concerned that the complexity of the migration is not really necessary. All those clalculations if we could just refund half.

Reply:
> All those clalculations if we could just refund half.

I think giving the runtime developers an example of how they could properly address issues with inconsistent deposit amounts is a good thing. We show it in the storage meter logic, but things are different there because it has the amounts related to item_added and bytes_added, which is not the case in the migrations. We only have the total deposit in the migrations. Returning "roughly half" seems as a vicious practice, which only multiplies inconsistency. And we basically have all the info needed to make proper calculations of refunds to be done pro-rata to cleared storage.

Reply:
Meanwhile removed the iterating over `PristineCode` from the migration, which has improved its weights as follows:

  | migration |     ref_time | % of v12 | proof_size | % of v12 |
  |-----------|--------------|----------|------------|----------|
  | v9        | 174281397290 |     0.20 |   27728976 |     0.40 |
  | v10       | 573395138290 |     0.66 |   41084911 |     0.60 |
  | v11       | 574898592340 |     0.66 |   41092187 |     0.60 |
  | v12       | 872520680768 |     1.00 |   68924789 |     1.00 |

, making it just 33-40% more costly than the previous ones.